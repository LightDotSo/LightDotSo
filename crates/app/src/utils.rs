// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use std::path::PathBuf;

use crate::constants::{DEV_APP_DIR, LIGHTDOTSO_APP_DIR};

// From: https://github.com/tomheaton/tauri-rspc-prisma/blob/e135a252a7c08d4a81847934ed73296c998f2753/core/src/utils.rs#L8
// License: MIT
pub fn get_lightdotso_dir() -> PathBuf {
    let path = platform_dirs::AppDirs::new(Some(&LIGHTDOTSO_APP_DIR.to_string()), true).unwrap();
    let mut data_dir = path.data_dir;

    // check if in dev mode
    if cfg!(debug_assertions) {
        data_dir.push(DEV_APP_DIR.to_string());
    }

    data_dir
}
