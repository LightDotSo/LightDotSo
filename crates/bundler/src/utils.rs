// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use ethers::types::{Address, U256};
use silius_uopool::Overhead;
/// Entire file is copied from https://github.com/Vid201/silius/blob/bc8b7b0039c9a2b02256fefc7eed3b2efc94bf96/bin/silius/src/utils.rs
/// License: MIT or Apache-2.0
use std::str::FromStr;

/// Parses address from string
pub fn parse_address(s: &str) -> Result<Address, String> {
    Address::from_str(s).map_err(|_| format!("String {s} is not a valid address"))
}

/// Parses U256 from string
pub fn parse_u256(s: &str) -> Result<U256, String> {
    U256::from_str_radix(s, 10).map_err(|_| format!("String {s} is not a valid U256"))
}

/// Helper function to calculate the call gas limit of a [UserOperation](UserOperation)
/// The function is invoked by the
/// [estimate_user_operation_gas](crates::uopool::estimate::estimate_user_operation_gas) method.
///
/// # Arguments
/// `paid` - The paid gas
/// `pre_op_gas` - The pre-operation gas
/// `fee_per_gas` - The fee per gas
///
/// # Returns
/// The call gas limit of the [UserOperation](UserOperation)
pub fn calculate_call_gas_limit(paid: U256, pre_op_gas: U256, fee_per_gas: U256) -> U256 {
    paid / fee_per_gas - pre_op_gas + Overhead::default().fixed
}
