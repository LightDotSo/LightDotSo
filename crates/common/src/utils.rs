// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use alloy::hex;
use eyre::{eyre, Result};

pub fn hex_to_bytes(hex_str: &str) -> Result<Vec<u8>, hex::FromHexError> {
    // Check if the hex string has the 0x prefix
    if !hex_str.starts_with("0x") {
        return Err(hex::FromHexError::InvalidStringLength);
    }

    // Remove the 0x prefix
    let pure_hex = &hex_str[2..];
    hex::decode(pure_hex) // Decode hex string to Vec<u8>
}

pub fn hex_to_bytes32(hex_str: &str) -> Result<[u8; 32]> {
    if !hex_str.starts_with("0x") {
        return Err(eyre!("Invalid hex string: no 0x prefix"));
    }

    let vec = hex_to_bytes(hex_str)?;

    if vec.len() != 32 {
        return Err(eyre!("Length mismatch: Vec is not 32 bytes long"));
    }

    let mut array = [0u8; 32];
    array.copy_from_slice(&vec);
    Ok(array)
}

// -----------------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_valid_hex() {
        let hex_str = "0x4a5b";
        let bytes = hex_to_bytes(hex_str).unwrap();
        assert_eq!(bytes, vec![74, 91]);
    }

    #[test]
    fn test_invalid_hex() {
        let hex_str = "0x4g";
        let result = hex_to_bytes(hex_str);
        assert!(result.is_err());
    }

    #[test]
    fn test_no_0x_prefix() {
        let hex_str = "4a5b";
        let result = hex_to_bytes(hex_str);
        assert!(result.is_err());
    }
}
