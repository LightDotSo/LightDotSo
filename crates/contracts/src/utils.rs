// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Rundler is free software: you can redistribute it and/or modify it under the
// terms of the GNU Lesser General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later version.
//
// Rundler is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with Rundler.
// If not, see https://www.gnu.org/licenses/.

#![allow(clippy::expect_used)]
#![allow(unused_doc_comments)]

use crate::entrypoint::EntryPoint::{ExecutionResult, FailedOp};
use alloy::{
    primitives::Bytes,
    providers::Provider,
    sol,
    sol_types::{ContractError, SolError},
};

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

/// Gets the revert data from a contract error if it is a revert error,
/// otherwise returns the original error.
/// From: https://github.com/alchemyplatform/rundler/blob/ae615d0faa97b61a7e0a3d0a21793f383560ae35/crates/utils/src/eth.rs#L31-37
/// License: GNU Lesser General Public License v3.0
pub fn get_revert_bytes<M: Provider>(error: ContractError<M>) -> Result<Bytes, ContractError<M>> {
    if let ContractError::Revert(revert) = error {
        Ok(Bytes::from(revert.reason().as_bytes().to_vec()))
    } else {
        Err(error)
    }
}

/// The abi for what happens when you just `revert("message")` in a contract
/// From: https://github.com/alchemyplatform/rundler/blob/ae615d0faa97b61a7e0a3d0a21793f383560ae35/crates/utils/src/eth.rs#L39-45
/// License: GNU Lesser General Public License v3.0
sol! {
  error Error(string);
}

/// Decode the revert data into an `ExecutionResult` or a `FailedOp`.
/// From: https://github.com/alchemyplatform/rundler/blob/ae615d0faa97b61a7e0a3d0a21793f383560ae35/crates/provider/src/ethers/entry_point.rs#L141-155
/// License: GNU Lesser General Public License v3.0
pub fn decode_simulate_handle_ops_revert(revert_data: Bytes) -> Result<ExecutionResult, String> {
    if let Ok(result) = ExecutionResult::abi_decode(&revert_data, true) {
        Ok(result)
    } else if let Ok(failed_op) = FailedOp::abi_decode(&revert_data, true) {
        Err(failed_op.reason)
    } else if let Ok(err) = Error::abi_decode(&revert_data, true) {
        Err(err._0)
    } else {
        Err(String::new())
    }
}
