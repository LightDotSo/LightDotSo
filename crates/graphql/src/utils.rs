// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use crate::constants::THE_GRAPH_HOSTED_SERVICE_URLS;
use eyre::{eyre, Result};

/// A helper function to get the graphql url for a given chain id.
pub fn get_graphql_url(chain_id: u64) -> Result<String> {
    let url = THE_GRAPH_HOSTED_SERVICE_URLS
        .get(&chain_id)
        .ok_or_else(|| eyre!("Chain id {} not supported", chain_id))?;
    Ok(url.to_string())
}
