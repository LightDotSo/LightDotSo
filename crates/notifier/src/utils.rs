// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::types::{NotificationOperation, UserOnlyNotification, WalletOnlyNotification};
use lightdotso_prisma::{ActivityEntity, ActivityOperation};
use serde_json::Value;

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

// Utility function to match the activity entity/operation combination
pub fn match_user_only_notification_with_activity(
    entity: &ActivityEntity,
    operation: &ActivityOperation,
    log: &Value,
) -> Option<UserOnlyNotification> {
    match (entity, operation) {
        (ActivityEntity::InviteCode, ActivityOperation::Update) => {
            // Check if the key of log is status is `USED`
            if log["status"].as_str() == Some("USED") {
                Some(UserOnlyNotification::InviteCodeAccepted)
            } else {
                None
            }
        }
        _ => None,
    }
}

// Utility function to match the activity entity/operation combination
pub fn match_wallet_only_notification_with_activity(
    entity: &ActivityEntity,
    operation: &ActivityOperation,
    log: &Value,
) -> Option<WalletOnlyNotification> {
    match (entity, operation) {
        (ActivityEntity::UserOperation, ActivityOperation::Create) => {
            Some(WalletOnlyNotification::UserOperationCreated)
        }
        (ActivityEntity::UserOperation, ActivityOperation::Update) => {
            if log["status"].as_str() == Some("EXECUTED") ||
                log["status"].as_str() == Some("REVERTED")
            {
                Some(WalletOnlyNotification::TransactionWithUserOperationExecuted)
            } else {
                None
            }
        }
        (ActivityEntity::Transaction, ActivityOperation::Update) => None,
        _ => None,
    }
}

// Combination utility function to match the activity entity/operation combination
pub fn match_notification_operation_with_activity(
    entity: &ActivityEntity,
    operation: &ActivityOperation,
    log: &Value,
) -> Option<NotificationOperation> {
    if let Some(res) = match_user_only_notification_with_activity(entity, operation, log) {
        return Some(NotificationOperation::UserOnly(res));
    }

    match_wallet_only_notification_with_activity(entity, operation, log)
        .map(NotificationOperation::WalletOnly)
}

// -----------------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use std::str::FromStr;

    #[test]
    fn test_match_user_only_notification_with_activity() {
        let entity = ActivityEntity::InviteCode;
        let operation = ActivityOperation::Update;
        let log = serde_json::Value::from_str(r#"{"status": "USED"}"#).unwrap();
        let result = match_user_only_notification_with_activity(&entity, &operation, &log);
        assert_eq!(result, Some(UserOnlyNotification::InviteCodeAccepted));
    }

    #[test]
    fn test_match_wallet_only_notification_with_activity() {
        let entity = ActivityEntity::UserOperation;
        let operation = ActivityOperation::Create;
        let result =
            match_wallet_only_notification_with_activity(&entity, &operation, &Value::Null);
        assert_eq!(result, Some(WalletOnlyNotification::UserOperationCreated));
    }

    #[test]
    fn test_match_notification_operation_with_activity() {
        let entity = ActivityEntity::UserOperation;
        let operation = ActivityOperation::Create;
        let result = match_notification_operation_with_activity(&entity, &operation, &Value::Null);
        assert_eq!(
            result,
            Some(NotificationOperation::WalletOnly(WalletOnlyNotification::UserOperationCreated))
        );
    }
}
