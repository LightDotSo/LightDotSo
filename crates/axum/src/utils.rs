// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use ethers_main::utils::hex;
use eyre::{eyre, Result};

pub fn hex_to_bytes(hex_str: &str) -> Result<Vec<u8>, hex::FromHexError> {
    // Check if the hex string has the 0x prefix
    if !hex_str.starts_with("0x") {
        return Err(hex::FromHexError::InvalidStringLength);
    }

    // Remove the 0x prefix
    let pure_hex = &hex_str[2..];
    hex::decode(pure_hex) // Decode hex string to Vec<u8>
}

pub fn hex_to_bytes32(hex_str: &str) -> Result<[u8; 32]> {
    if !hex_str.starts_with("0x") {
        return Err(eyre!("Invalid hex string: no 0x prefix"));
    }

    let vec = hex_to_bytes(hex_str)?;

    if vec.len() != 32 {
        return Err(eyre!("Length mismatch: Vec is not 32 bytes long"));
    }

    let mut array = [0u8; 32];
    array.copy_from_slice(&vec);
    Ok(array)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_valid_hex() {
        let hex_str = "0x4a5b";
        let bytes = hex_to_bytes(hex_str).unwrap();
        assert_eq!(bytes, vec![74, 91]);
    }

    #[test]
    fn test_invalid_hex() {
        let hex_str = "0x4g";
        let result = hex_to_bytes(hex_str);
        assert!(result.is_err());
    }

    #[test]
    fn test_no_0x_prefix() {
        let hex_str = "4a5b";
        let result = hex_to_bytes(hex_str);
        assert!(result.is_err());
    }
}
