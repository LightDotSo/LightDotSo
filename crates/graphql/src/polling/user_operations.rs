// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use crate::utils::get_graphql_url;
use eyre::Result;

#[cynic::schema("graph")]
mod schema {}

#[derive(cynic::QueryVariables, Debug)]
pub struct GetUserOperationsQueryVariables {
    pub min_block: BigInt,
    pub min_index: BigInt,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "GetUserOperationsQueryVariables")]
pub struct GetUserOperationsQuery {
    #[arguments(first: 300, where: { blockNumber_gt: $min_block, index_gt: $min_index }, orderBy: "index")]
    pub user_operations: Vec<UserOperation>,
    pub _meta: Option<Meta>,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "_Meta_")]
pub struct Meta {
    pub block: Block,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "_Block_")]
pub struct Block {
    pub number: i32,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct UserOperation {
    pub id: Bytes,
    pub index: BigInt,
    pub sender: Bytes,
    pub nonce: BigInt,
    pub init_code: Bytes,
    pub call_data: Bytes,
    pub call_gas_limit: BigInt,
    pub verification_gas_limit: BigInt,
    pub pre_verification_gas: BigInt,
    pub max_fee_per_gas: BigInt,
    pub max_priority_fee_per_gas: BigInt,
    pub paymaster_and_data: Bytes,
    pub signature: Bytes,
    pub block_number: BigInt,
    pub block_timestamp: BigInt,
    pub transaction_hash: Bytes,
    pub entry_point: Bytes,
    pub user_operation_event: Option<UserOperationEvent>,
    pub user_operation_revert_reason: Option<UserOperationRevertReason>,
    pub light_wallet: LightWallet,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct UserOperationRevertReason {
    pub id: Bytes,
    pub index: BigInt,
    pub user_op_hash: Bytes,
    pub sender: Bytes,
    pub nonce: BigInt,
    pub revert_reason: Bytes,
    pub transaction_hash: Bytes,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct UserOperationEvent {
    pub id: Bytes,
    pub index: BigInt,
    pub user_op_hash: Bytes,
    pub paymaster: Bytes,
    pub nonce: BigInt,
    pub transaction_hash: Bytes,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct LightWallet {
    pub id: Bytes,
    pub index: BigInt,
    pub address: Bytes,
    pub image_hash: Option<Bytes>,
    pub user_op_hash: Bytes,
    pub sender: Bytes,
    pub factory: Bytes,
    pub paymaster: Bytes,
    pub block_number: BigInt,
    pub block_timestamp: BigInt,
    pub transaction_hash: Bytes,
}

#[derive(cynic::Scalar, Debug, Clone)]
pub struct BigInt(pub String);

#[derive(cynic::Scalar, Debug, Clone)]
pub struct Bytes(pub String);

pub fn build_user_operations_query(
    vars: GetUserOperationsQueryVariables,
) -> cynic::Operation<GetUserOperationsQuery, GetUserOperationsQueryVariables> {
    use cynic::QueryBuilder;

    GetUserOperationsQuery::build(vars)
}

pub fn run_user_operations_query(
    chain_id: u64,
    vars: GetUserOperationsQueryVariables,
) -> Result<cynic::GraphQlResponse<GetUserOperationsQuery>> {
    use cynic::http::ReqwestBlockingExt;

    let query = build_user_operations_query(vars);

    Ok(reqwest::blocking::Client::new().post(get_graphql_url(chain_id)?).run_graphql(query)?)
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn snapshot_test_query() {
        // Running a snapshot test of the query building functionality as that gives us
        // a place to copy and paste the actual GQL we're using for running elsewhere,
        // and also helps ensure we don't change queries by mistake

        let query = build_user_operations_query(GetUserOperationsQueryVariables {
            min_block: BigInt("0".to_string()),
            min_index: BigInt("0".to_string()),
        });

        insta::assert_snapshot!(query.query);
    }

    #[test]
    fn test_running_query() {
        let result = run_user_operations_query(
            1,
            GetUserOperationsQueryVariables {
                min_block: BigInt("0".to_string()),
                min_index: BigInt("0".to_string()),
            },
        )
        .unwrap();
        if result.errors.is_some() {
            assert_eq!(result.errors.unwrap().len(), 0);
        }
    }
}
