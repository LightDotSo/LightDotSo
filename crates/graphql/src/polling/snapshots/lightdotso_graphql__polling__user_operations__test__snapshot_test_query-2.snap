---
source: crates/graphql/src/polling/user_operations.rs
expression: query.query
---
query GetUserOperationQuery($id: String!) {
  userOperation(id: $id) {
    id
    index
    sender
    nonce
    initCode
    callData
    callGasLimit
    verificationGasLimit
    preVerificationGas
    maxFeePerGas
    maxPriorityFeePerGas
    paymasterAndData
    signature
    blockNumber
    blockTimestamp
    transactionHash
    entryPoint
    paymaster
    userOperationEvent {
      id
      index
      userOpHash
      paymaster
      nonce
      transactionHash
    }
    userOperationRevertReason {
      id
      index
      userOpHash
      sender
      nonce
      revertReason
      transactionHash
    }
    lightWallet {
      id
      index
      address
      imageHash
      userOpHash
      sender
      factory
      paymaster
      blockNumber
      blockTimestamp
      transactionHash
    }
    transaction {
      id
      hash
      index
      from
      to
      value
      gasLimit
      gasPrice
      input
      nonce
      receipt {
        id
        transactionHash
        transactionIndex
        blockHash
        blockNumber
        cumulativeGasUsed
        gasUsed
        status
        logsBloom
        logs {
          id
          address
          topics
          data
          blockHash
          blockNumber
          transactionHash
          transactionIndex
          logIndex
        }
      }
    }
    logs {
      id
      address
      topics
      data
      blockHash
      blockNumber
      transactionHash
      transactionIndex
      logIndex
    }
  }
  _meta {
    block {
      number
    }
  }
}


